File: scripts/algorithms/G/Guess Number Higher or Lower II/Guess Number Higher or Lower II.cpp
// Runtime: 226 ms (Top 55.81%) | Memory: 8.2 MB (Top 44.55%)
class Solution {
public:
    vector<vector<int>> dp;

    int solve(int start, int end)
    {
        if(start>= end)
            return 0;

        if(dp[start][end] != -1)
            return dp[start][end];

        int ans = 0;
        int result = INT_MAX;
        for(int i=start; i<=end; i++)
        {
            int left = solve(start,i-1);
            int right = solve(i+1,end);
            ans = max(left,right) + i; // this line gurantee to include the money that is needed to win higher values
            result = min(ans,result);
        }

        return dp[start][end] = result;
    }

    int getMoneyAmount(int n) {
        int ans = 0;
        dp = vector<vector<int>>(n+1,vector<int>(n+1,-1));
        return solve(1,n);
    }
};

File: scripts/algorithms/G/Guess Number Higher or Lower II/Guess Number Higher or Lower II.java
class Solution {
    public int getMoneyAmount(int n) {
        int dp[][]=new int[n+1][n+1];
        for(int a[]:dp){
            Arrays.fill(a,-1);
        }
        return solve(1,n,dp);
    }
    static int solve(int start,int end,int[][] dp){
    if(start>=end) return 0;
    if(dp[start][end]!=-1) return dp[start][end];
    int min=Integer.MAX_VALUE;
    for(int i=start;i<=end;i++){
    min=Math.min(min,i+Math.max(solve(start,i-1,dp),solve(i+1,end,dp)));
    }
    dp[start][end]=min;
    return min;
    }
}